### PR TITLE
Apply quantization on megablox kernel; support both training and serving

### PR DESCRIPTION
**Pull request by lenscloth**
# Description

Support quantization on megablox for both training and serving

- Tested correctness after apply quantization on megablox.
- This is the squashed version of https://github.com/AI-Hypercomputer/maxtext/pull/1062

# Tests

End to end tests on mixtral-8x22b model with max engine.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by gobbleturk — APPROVED** [review]

File: MaxText/layers/linears.py [hunk 1]
```diff
@@ -386,15 +387,14 @@ def permute(self, inputs, gate_logits):
...
+  def unpermute(self, intermediate, sorted_selected_experts, weights, batch_size: int, sequence_length: int):
     """Unpermute tokens to original order and combine weights."""
 
     unsort_intermediate = jnp.take(intermediate, indices=jnp.argsort(sorted_selected_experts), axis=0)
     reshaped_weights = jnp.reshape(weights, (-1, self.num_experts_per_tok))
-    tensor_parallelism = self.config.ici_tensor_parallelism * self.config.dcn_tensor_parallelism
     reshaped_intermediate = jnp.reshape(
         unsort_intermediate,
-        (-1, self.num_experts_per_tok, self.config.emb_dim // tensor_parallelism),
+        (reshaped_weights.shape[0], self.num_experts_per_tok, -1),
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:397]: Discussed in the group chat.

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -404,7 +404,7 @@ def unpermute(self, intermediate, sorted_selected_experts, weights):
           reshaped_weights.astype(jnp.float32),
           precision=matmul_precision,
       )
-    return output.reshape(-1, self.config.max_target_length, self.config.emb_dim // tensor_parallelism).astype(self.dtype)
+    return output.reshape(batch_size, sequence_length, -1).astype(self.dtype)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:407]: Discussed in the group chat.

File: MaxText/layers/linears.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lenscloth) [reply MaxText/layers/linears.py:397]:
```
with jax.named_scope("weight_sum"):
      matmul_precision = lax.Precision(self.config.matmul_precision)
      output = jnp.einsum(
          "BKE,BK -> BE",
          reshaped_intermediate.astype(jnp.float32),
          reshaped_weights.astype(jnp.float32),
          precision=matmul_precision,
      )
 ```
 
 From the einsum eq below, batch size can be referenced from reshaped_weight.

**Review by RissyRan — APPROVED** [review]
Thank you! Others LGTM

File: MaxText/layers/linears.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lenscloth) [reply MaxText/layers/linears.py:407]: The shape before permute / unpermute should be same. I believe we can infer this from batch size and sequence length.

File: MaxText/layers/linears.py [hunk 3]
```diff
@@ -692,11 +726,14 @@ def __call__(self, inputs):
         name="gate",
         matmul_precision=self.config.matmul_precision,
     )(inputs)
-
+    cfg = self.config
```
[the remark this answers is not in the document]
Comment (lenscloth) [reply MaxText/layers/linears.py]: Done

**Pull request merged**